text/0000-relaxed-dst-field-ordering.md

```diff
@@ -0,0 +1,65 @@
+- Feature Name: `relaxed_dst_field_ordering`
+- Start Date: 2024-10-18
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Relax the requirements on struct field ordering for dynamically sized fields:
+
+* for `repr(Rust)`, `?Sized` fields can be anywhere in the field list, as long as there is only one
+* for `repr(C)`, `?Sized` fields only have to be the last non-ZST field, and can be followed by ZST fields
+* for `repr(transparent)`, apply both rules, since only one non-ZST field is allowed anyway
+
+# Motivation
+[motivation]: #motivation
+
+Rust allows creating structs with dynamically sized fields, but in a very limited way: since the size is dynamic, dynamically sized fields must be the last field in the struct, to avoid making the offsets of fields also dynamic.
+
+However, `repr(Rust)` allows reordering fields, and this is not not reflected in this rule for dynamically sized fields: no matter what you do, the dynamically sized field must be at the end. This is inconsistent with the rest of the language and limits the ability of authors to reorder the fields in a more natural way, like they can for statically sized structs.
+
+Additionally, Rust has fully committed to zero-sized fields being truly invisible to struct layout, encoding this in the definition of `repr(transparent)`. So, why are dynamically sized fields unable to be followed by zero-sized fields, or reordered among statically sized fields in a `repr(Rust)` struct?
```

Comment:
pretty sure the *alignment* of a ZST is part of the layout and is visible (and therefore [E0690](https://doc.rust-lang.org/error_codes/E0690.html) "transparent struct needs at most one field with non-trivial size or alignment" exists)

given that the Ref-level explanation assumed `#[repr(transparent)]` and `#[repr(C)]` will be using the same logic i think you just need to mention "ZST" in this RFC is restricted to (size 0, align 1) only.

Reply:
Hopefully the reworded explanation is better: the new rules apply to both `repr(Rust)` and `repr(transparent)`, but the restrictions on what ZSTs are allowed for `repr(transparent)` remain.